File: examples/agentic-mentorloop-py/agentic_mentorloop/tools/csv_pulse.py
import csv
import os
from datetime import datetime
from typing import Dict, List, Optional, Union


class EngagementPulse:
    def __init__(self, data_path: str = None):
        if data_path is None:
            data_path = os.path.join(os.path.dirname(__file__), '../../../data')
        self.data_path = data_path
    
    def run(self, options: Optional[Dict] = None) -> Dict:
        try:
            mode = options.get('mode', 'summary') if options else 'summary'
            limit = options.get('limit', 5) if options else 5
            
            checkins = self._load_checkins()
            messages = self._load_messages()
            users = self._load_users()
            pairings = self._load_pairings()
            
            now = datetime.now()
            pair_stats = {}
            
            # Analyse checkins
            for checkin in checkins:
                pair_id = checkin['pair_id']
                checkin_date = datetime.fromisoformat(checkin['timestamp'].replace('Z', '+00:00'))
                
                if pair_id not in pair_stats or checkin_date > pair_stats[pair_id]['last_checkin']:
                    days_since = (now - checkin_date.replace(tzinfo=None)).days
                    pair_stats[pair_id] = {
                        'last_checkin': checkin_date,
                        'balance': 'balanced',
                        'days_since': days_since
                    }
            
            # Analyse message balance for last 50 messages per pair
            pair_messages = {}
            for msg in messages:
                pair_id = msg['pair_id']
                if pair_id not in pair_messages:
                    pair_messages[pair_id] = []
                pair_messages[pair_id].append(msg)
            
            for pair_id, msgs in pair_messages.items():
                recent_messages = msgs[-50:]
                mentor_count = sum(1 for m in recent_messages if m['author_role'] == 'mentor')
                mentee_count = sum(1 for m in recent_messages if m['author_role'] == 'mentee')
                
                balance = 'balanced'
                if mentor_count > mentee_count * 1.5:
                    balance = 'mentor_heavy'
                elif mentee_count > mentor_count * 1.5:
                    balance = 'mentee_heavy'
                
                if pair_id in pair_stats:
                    pair_stats[pair_id]['balance'] = balance
            
            # Find dormant pairs (> 14 days)
            dormant_pairs = []
            for pair_id, stats in pair_stats.items():
                if stats['days_since'] > 14:
                    dormant_pairs.append({'pair_id': pair_id, 'days_since': stats['days_since']})
            
            if mode == 'list':
                # Sort dormant pairs by days since last checkin (descending)
                dormant_pairs.sort(key=lambda x: x['days_since'], reverse=True)
                
                # Map to mentee information
                dormant_mentees = []
                for item in dormant_pairs[:limit]:
                    pair_id = item['pair_id']
                    days_since = item['days_since']
                    
                    pairing = next((p for p in pairings if p['pair_id'] == pair_id), None)
                    if not pairing:
                        dormant_mentees.append({
                            'mentee_id': 'unknown',
                            'first_name': 'Unknown',
                            'pair_id': pair_id,
                            'last_checkin_days': days_since
                        })
                        continue
                    
                    mentee = next((u for u in users if u['user_id'] == pairing['mentee_id']), None)
                    dormant_mentees.append({
                        'mentee_id': pairing['mentee_id'],
                        'first_name': mentee['first_name'] if mentee else 'Unknown',
                        'pair_id': pair_id,
                        'last_checkin_days': days_since
                    })
                
                return {
                    'type': 'list',
                    'dormant_count': len(dormant_pairs),
                    'mentees': dormant_mentees
                }
            else:
                # Summary mode
                max_days = 0
                for stats in pair_stats.values():
                    if stats['days_since'] > max_days:
                        max_days = stats['days_since']
                
                # Get overall balance
                balances = [s['balance'] for s in pair_stats.values()]
                mentor_heavy = sum(1 for b in balances if b == 'mentor_heavy')
                mentee_heavy = sum(1 for b in balances if b == 'mentee_heavy')
                
                if mentor_heavy > len(balances) / 2:
                    overall_balance = 'mentor_heavy'
                elif mentee_heavy > len(balances) / 2:
                    overall_balance = 'mentee_heavy'
                else:
                    overall_balance = 'balanced'
                
                return {
                    'type': 'summary',
                    'sample': len(pair_stats),
                    'dormant': len(dormant_pairs),
                    'balance': overall_balance,
                    'last_checkin_days': max_days
                }
        except Exception as e:
            return {
                'type': 'error',
                'message': str(e),
                'hint': 'Check that data files exist and are properly formatted'
            }
    
    def _load_checkins(self) -> List[Dict]:
        file_path = os.path.join(self.data_path, 'checkins.csv')
        checkins = []
        with open(file_path, 'r') as f:
            reader = csv.DictReader(f)
            for row in reader:
                checkins.append(row)
        return checkins
    
    def _load_messages(self) -> List[Dict]:
        file_path = os.path.join(self.data_path, 'messages.csv')
        messages = []
        with open(file_path, 'r') as f:
            reader = csv.DictReader(f)
            for row in reader:
                messages.append(row)
        return messages
    
    def _load_users(self) -> List[Dict]:
        file_path = os.path.join(self.data_path, 'users.csv')
        users = []
        with open(file_path, 'r') as f:
            reader = csv.DictReader(f)
            for row in reader:
                users.append(row)
        return users
    
    def _load_pairings(self) -> List[Dict]:
        file_path = os.path.join(self.data_path, 'pairings.csv')
        pairings = []
        with open(file_path, 'r') as f:
            reader = csv.DictReader(f)
            for row in reader:
                pairings.append(row)
        return pairings

File: examples/agentic-mentorloop-ts/src/tools/csvPulse.ts
import * as fs from 'fs';
import * as path from 'path';

interface CheckIn {
  pair_id: string;
  timestamp: string;
  mentee_score: number;
  mentor_score: number;
  notes: string;
}

interface Message {
  pair_id: string;
  timestamp: string;
  author_role: string;
  channel: string;
  text: string;
}

interface User {
  user_id: string;
  role: string;
  email: string;
  timezone: string;
  first_name: string;
  joined_at: string;
}

interface Pairing {
  pair_id: string;
  mentor_id: string;
  mentee_id: string;
  programme_id: string;
  started_at: string;
}

interface EngagementOptions {
  mode?: 'summary' | 'list';
  limit?: number;
  prompt?: string;
}

interface SummaryResult {
  type: 'summary';
  sample: number;
  dormant: number;
  balance: string;
  last_checkin_days: number;
}

interface ListResult {
  type: 'list';
  dormant_count: number;
  mentees: Array<{
    mentee_id: string;
    first_name: string;
    pair_id: string;
    last_checkin_days: number;
  }>;
}

interface ErrorResult {
  type: 'error';
  message: string;
  hint: string;
}

type EngagementResult = SummaryResult | ListResult | ErrorResult;

export class EngagementPulse {
  private dataPath: string;

  constructor(dataPath: string = path.join(__dirname, '../../../data')) {
    this.dataPath = dataPath;
  }

  run(options?: EngagementOptions): EngagementResult {
    try {
      const mode = options?.mode || 'summary';
      const limit = options?.limit || 5;
      
      const checkins = this.loadCheckins();
      const messages = this.loadMessages();
      const users = this.loadUsers();
      const pairings = this.loadPairings();
      
      const now = new Date();
      const pairStats = new Map<string, { lastCheckin: Date, balance: string, daysSince: number }>();
      
      // Analyse checkins
      checkins.forEach(checkin => {
        const pairId = checkin.pair_id;
        const checkinDate = new Date(checkin.timestamp);
        
        if (!pairStats.has(pairId) || checkinDate > pairStats.get(pairId)!.lastCheckin) {
          const daysSince = Math.floor((now.getTime() - checkinDate.getTime()) / (1000 * 60 * 60 * 24));
          pairStats.set(pairId, { lastCheckin: checkinDate, balance: 'balanced', daysSince });
        }
      });
      
      // Analyse message balance for last 50 messages per pair
      const pairMessages = new Map<string, Message[]>();
      messages.forEach(msg => {
        if (!pairMessages.has(msg.pair_id)) {
          pairMessages.set(msg.pair_id, []);
        }
        pairMessages.get(msg.pair_id)!.push(msg);
      });
      
      pairMessages.forEach((msgs, pairId) => {
        const recentMessages = msgs.slice(-50);
        const mentorCount = recentMessages.filter(m => m.author_role === 'mentor').length;
        const menteeCount = recentMessages.filter(m => m.author_role === 'mentee').length;
        
        let balance = 'balanced';
        if (mentorCount > menteeCount * 1.5) {
          balance = 'mentor_heavy';
        } else if (menteeCount > mentorCount * 1.5) {
          balance = 'mentee_heavy';
        }
        
        if (pairStats.has(pairId)) {
          pairStats.get(pairId)!.balance = balance;
        }
      });
      
      // Find dormant pairs (> 14 days)
      const dormantPairs: Array<{pairId: string, daysSince: number}> = [];
      pairStats.forEach((stats, pairId) => {
        if (stats.daysSince > 14) {
          dormantPairs.push({ pairId, daysSince: stats.daysSince });
        }
      });
      
      if (mode === 'list') {
        // Sort dormant pairs by days since last checkin (descending)
        dormantPairs.sort((a, b) => b.daysSince - a.daysSince);
        
        // Map to mentee information
        const dormantMentees = dormantPairs.slice(0, limit).map(({ pairId, daysSince }) => {
          const pairing = pairings.find(p => p.pair_id === pairId);
          if (!pairing) {
            return {
              mentee_id: 'unknown',
              first_name: 'Unknown',
              pair_id: pairId,
              last_checkin_days: daysSince
            };
          }
          
          const mentee = users.find(u => u.user_id === pairing.mentee_id);
          return {
            mentee_id: pairing.mentee_id,
            first_name: mentee?.first_name || 'Unknown',
            pair_id: pairId,
            last_checkin_days: daysSince
          };
        });
        
        return {
          type: 'list',
          dormant_count: dormantPairs.length,
          mentees: dormantMentees
        };
      } else {
        // Summary mode
        let maxDays = 0;
        pairStats.forEach(stats => {
          if (stats.daysSince > maxDays) maxDays = stats.daysSince;
        });
        
        // Get overall balance
        const balances = Array.from(pairStats.values()).map(s => s.balance);
        const overallBalance = balances.filter(b => b === 'mentor_heavy').length > balances.length / 2 
          ? 'mentor_heavy' 
          : balances.filter(b => b === 'mentee_heavy').length > balances.length / 2
          ? 'mentee_heavy'
          : 'balanced';
        
        return {
          type: 'summary',
          sample: pairStats.size,
          dormant: dormantPairs.length,
          balance: overallBalance,
          last_checkin_days: maxDays
        };
      }
    } catch (error) {
      return {
        type: 'error',
        message: error instanceof Error ? error.message : 'Unknown error occurred',
        hint: 'Check that data files exist and are properly formatted'
      };
    }
  }

  private loadCheckins(): CheckIn[] {
    const filePath = path.join(this.dataPath, 'checkins.csv');
    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.trim().split('\n');
    
    return lines.slice(1).map(line => {
      const values = line.split(',');
      return {
        pair_id: values[0],
        timestamp: values[1],
        mentee_score: parseInt(values[2]),
        mentor_score: parseInt(values[3]),
        notes: values[4]
      };
    });
  }

  private loadMessages(): Message[] {
    const filePath = path.join(this.dataPath, 'messages.csv');
    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.trim().split('\n');
    
    return lines.slice(1).map(line => {
      const values = line.split(',');
      return {
        pair_id: values[0],
        timestamp: values[1],
        author_role: values[2],
        channel: values[3],
        text: values.slice(4).join(',')
      };
    });
  }

  private loadUsers(): User[] {
    const filePath = path.join(this.dataPath, 'users.csv');
    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.trim().split('\n');
    
    return lines.slice(1).map(line => {
      const values = line.split(',');
      return {
        user_id: values[0],
        role: values[1],
        email: values[2],
        timezone: values[3],
        first_name: values[4],
        joined_at: values[5]
      };
    });
  }

  private loadPairings(): Pairing[] {
    const filePath = path.join(this.dataPath, 'pairings.csv');
    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.trim().split('\n');
    
    return lines.slice(1).map(line => {
      const values = line.split(',');
      return {
        pair_id: values[0],
        mentor_id: values[1],
        mentee_id: values[2],
        programme_id: values[3],
        started_at: values[4]
      };
    });
  }
}